File: modules/application/src/main/scala/com/streamfirst/iceberg/hybrid/application/WriteCoordinator.scala
package com.streamfirst.iceberg.hybrid.application

import com.streamfirst.iceberg.hybrid.domain.DomainError.{CatalogError, SyncError}
import com.streamfirst.iceberg.hybrid.domain.{
  CommitId,
  CommitRequest,
  DomainError,
  Region,
  TableId,
  TableMetadata,
  WriteJob,
  WriteJobId,
  WriteJobStatus
}
import com.streamfirst.iceberg.hybrid.ports.{
  CatalogPort,
  CommitGatePort,
  CommitStatus,
  RegistryPort,
  SyncPort
}
import zio.{IO, Ref, ZIO, ZLayer}

/** Coordinates write operations across regions. Handles the distributed commit protocol to ensure
  * consistency when writing to geo-distributed tables.
  */
final case class WriteCoordinator(
    commitGatePort: CommitGatePort,
    catalogPort: CatalogPort,
    syncPort: SyncPort,
    registryPort: RegistryPort,
    writeJobsRef: Ref[Map[WriteJobId, WriteJob]]
):
  /** Coordinates a write operation across all regions with job tracking. Implements the distributed commit protocol. */
  def coordinateWrite(
      tableId: TableId,
      metadata: TableMetadata,
      sourceRegion: Region
  ): IO[SyncError | CatalogError, WriteJobId] =
    val writeJob = WriteJob.create(tableId, sourceRegion)
    val request = CommitRequest.create(tableId, sourceRegion, metadata)
    
    for
      // Create and track the write job
      _ <- writeJobsRef.update(_.updated(writeJob.writeJobId, writeJob))
      _ <- ZIO.logInfo(s"Created write job ${writeJob.writeJobId} for table $tableId from region ${sourceRegion.id}")

      // Step 1: Request commit approval from global gate
      _ <- updateWriteJobStatus(writeJob.writeJobId, WriteJobStatus.RequestingApproval)
      approval <- commitGatePort.requestCommitApproval(request)
      targetRegions = approval.approvedRegions.filter(_ != sourceRegion)
      _ <- updateWriteJobTargetRegions(writeJob.writeJobId, targetRegions)
      _ <- updateWriteJobStatus(writeJob.writeJobId, WriteJobStatus.Approved)

      // Step 2: Commit to local catalog first
      _ <- updateWriteJobStatus(writeJob.writeJobId, WriteJobStatus.CommittingLocal)
      actualCommitId <- catalogPort.commitMetadata(metadata)
      _ <- updateWriteJobCommitId(writeJob.writeJobId, actualCommitId)

      // Step 3: Create sync events for all other regions
      _ <- updateWriteJobStatus(writeJob.writeJobId, WriteJobStatus.SynchronizingRegions)
      _ <- ZIO.foreachParDiscard(targetRegions) { targetRegion =>
        for
          metadataEvent <- syncPort.createMetadataSyncEvent(metadata, targetRegion)
          dataEvent <- syncPort.createDataSyncEvent(metadata, metadata.dataFiles, targetRegion)
          _ <- syncPort.publishSyncEvent(metadataEvent)
          _ <- syncPort.publishSyncEvent(dataEvent)
        yield ()
      }

      // Step 4: Notify commit gate of successful completion
      _ <- commitGatePort.notifyCommitCompleted(actualCommitId, sourceRegion)
      _ <- updateWriteJobStatus(writeJob.writeJobId, WriteJobStatus.Completed)

      _ <- ZIO.logInfo(s"Successfully coordinated write job ${writeJob.writeJobId} for table $tableId, commit $actualCommitId")
    yield writeJob.writeJobId

  /** Handles a failed write by notifying the commit gate and triggering cleanup. */
  def handleWriteFailure(commitId: CommitId, region: Region, error: String): IO[SyncError, Unit] =
    for
      _ <- ZIO.logError(s"Write failed for commit $commitId in region ${region.id}: $error")
      _ <- commitGatePort.notifyCommitFailed(commitId, region, error)
    // TODO: Implement rollback logic if needed
    yield ()

  /** Gets the current status of a write operation. */
  def getWriteStatus(commitId: CommitId): IO[SyncError, CommitStatus] =
    commitGatePort.getCommitStatus(commitId)

  /** Gets the write job by its ID. */
  def getWriteJob(writeJobId: WriteJobId): IO[SyncError, Option[WriteJob]] =
    writeJobsRef.get.map(_.get(writeJobId))

  /** Lists all active write jobs. */
  def listActiveWriteJobs(): IO[SyncError, List[WriteJob]] =
    writeJobsRef.get.map(_.values.filterNot(job => job.status match {
      case WriteJobStatus.Completed | WriteJobStatus.Failed(_) => true
      case _ => false
    }).toList)

  // Helper methods for updating write job state
  private def updateWriteJobStatus(writeJobId: WriteJobId, status: WriteJobStatus): IO[SyncError, Unit] =
    writeJobsRef.update { jobs =>
      jobs.get(writeJobId) match {
        case Some(job) => jobs.updated(writeJobId, job.withStatus(status))
        case None => jobs
      }
    }

  private def updateWriteJobCommitId(writeJobId: WriteJobId, commitId: CommitId): IO[SyncError, Unit] =
    writeJobsRef.update { jobs =>
      jobs.get(writeJobId) match {
        case Some(job) => jobs.updated(writeJobId, job.withCommitId(commitId))
        case None => jobs
      }
    }

  private def updateWriteJobTargetRegions(writeJobId: WriteJobId, regions: List[Region]): IO[SyncError, Unit] =
    writeJobsRef.update { jobs =>
      jobs.get(writeJobId) match {
        case Some(job) => jobs.updated(writeJobId, job.withTargetRegions(regions))
        case None => jobs
      }
    }

object WriteCoordinator:
  val live
      : ZLayer[CommitGatePort & CatalogPort & SyncPort & RegistryPort, Nothing, WriteCoordinator] =
    ZLayer {
      for
        commitGatePort <- ZIO.service[CommitGatePort]
        catalogPort <- ZIO.service[CatalogPort]
        syncPort <- ZIO.service[SyncPort]
        registryPort <- ZIO.service[RegistryPort]
        writeJobsRef <- Ref.make(Map.empty[WriteJobId, WriteJob])
      yield WriteCoordinator(commitGatePort, catalogPort, syncPort, registryPort, writeJobsRef)
    }

File: modules/application/src/main/java/com/streamfirst/iceberg/hybrid/application/WriteCoordinator.java
package com.streamfirst.iceberg.hybrid.application;

import com.streamfirst.iceberg.hybrid.domain.*;
import com.streamfirst.iceberg.hybrid.ports.*;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;

import java.time.Instant;
import java.util.List;
import java.util.UUID;
import java.util.concurrent.CompletableFuture;

/**
 * Coordinates write operations across the geo-distributed system.
 * Implements the distributed commit protocol ensuring consistency across regions.
 * Handles the complete write workflow from commit request to global replication.
 */
@Slf4j
@RequiredArgsConstructor
public class WriteCoordinator {
    
    private final CatalogPort catalogPort;
    private final CommitGatePort commitGatePort;
    private final SyncPort syncPort;
    private final RegistryPort registryPort;

    /**
     * Executes a write operation with geo-distributed consistency.
     * 
     * @param request the commit request containing write details
     * @return future that completes when write is committed globally
     */
    public CompletableFuture<CommitId> executeWrite(CommitRequest request) {
        log.info("Starting write coordination for {}", request);
        
        return requestCommitApproval(request)
            .thenCompose(this::validateApprovalResult)
            .thenCompose(approvalResult -> commitToGlobalCatalog(request))
            .thenCompose(commitId -> triggerReplication(request, commitId))
            .whenComplete((commitId, throwable) -> {
                if (throwable != null) {
                    log.error("Write coordination failed for {}", request, throwable);
                    commitGatePort.releaseCommitLock(request);
                } else {
                    log.info("Write coordination completed successfully for {} with commit {}", 
                             request, commitId);
                }
            });
    }

    /**
     * Requests approval from all required regions for the commit.
     */
    private CompletableFuture<Result<String>> requestCommitApproval(CommitRequest request) {
        List<Region> requiredRegions = commitGatePort.getRequiredApprovalRegions(request.getTableId());
        log.debug("Requesting approval from {} regions for {}", requiredRegions.size(), request);
        
        return commitGatePort.requestCommitApproval(request);
    }

    /**
     * Validates that the approval result indicates success.
     */
    private CompletableFuture<Result<String>> validateApprovalResult(Result<String> result) {
        if (result.isSuccess()) {
            return CompletableFuture.completedFuture(result);
        } else {
            return CompletableFuture.failedFuture(
                new IllegalStateException("Commit approval failed: " + result.getErrorMessage().orElse("Unknown error")));
        }
    }

    /**
     * Commits the metadata to the global catalog once approval is received.
     */
    private CompletableFuture<CommitId> commitToGlobalCatalog(CommitRequest request) {
        try {
            CommitId commitId = catalogPort.commitMetadata(request);
            log.debug("Successfully committed metadata to global catalog: {}", commitId);
            return CompletableFuture.completedFuture(commitId);
        } catch (Exception e) {
            log.error("Failed to commit metadata to global catalog for {}", request, e);
            return CompletableFuture.failedFuture(e);
        }
    }

    /**
     * Triggers replication of the committed changes to all regions.
     */
    private CompletableFuture<CommitId> triggerReplication(CommitRequest request, CommitId commitId) {
        List<Region> targetRegions = registryPort.getTableRegions(request.getTableId());
        
        // Create metadata sync events for all regions except source
        targetRegions.stream()
            .filter(region -> !region.equals(request.getSourceRegion()))
            .forEach(targetRegion -> {
                try {
                    TableMetadata metadata = new TableMetadata(
                        request.getTableId(),
                        commitId,
                        request.getSourceRegion(),
                        request.getRequestTime(),
                        request.getNewDataFiles(),
                        request.getUpdatedSchema()
                    );
                    
                    SyncEvent metadataEvent = syncPort.createMetadataSyncEvent(metadata, targetRegion);
                    syncPort.publishSyncEvent(metadataEvent);
                    
                    if (!request.getNewDataFiles().isEmpty()) {
                        SyncEvent dataEvent = syncPort.createDataSyncEvent(
                            metadata, request.getNewDataFiles(), targetRegion);
                        syncPort.publishSyncEvent(dataEvent);
                    }
                    
                    log.debug("Created sync events for region {} and commit {}", 
                             targetRegion, commitId);
                } catch (Exception e) {
                    log.error("Failed to create sync events for region {} and commit {}", 
                             targetRegion, commitId, e);
                }
            });

        // Release the commit lock
        commitGatePort.releaseCommitLock(request);
        
        return CompletableFuture.completedFuture(commitId);
    }

    /**
     * Creates a new table in the catalog and registers it across regions.
     */
    public void createTable(TableId tableId, String schema, Region sourceRegion) {
        log.info("Creating table {} in region {}", tableId, sourceRegion);
        
        try {
            catalogPort.createTable(tableId, schema, sourceRegion);
            
            // Register table location in the source region
            String dataPath = generateTableDataPath(tableId);
            registryPort.registerTableLocation(tableId, sourceRegion, dataPath);
            
            log.info("Successfully created table {} in region {}", tableId, sourceRegion);
        } catch (Exception e) {
            log.error("Failed to create table {} in region {}", tableId, sourceRegion, e);
            throw e;
        }
    }

    /**
     * Generates a standard data path for a new table.
     */
    private String generateTableDataPath(TableId tableId) {
        return String.format("/data/%s/%s/%s", 
                           tableId.namespace(), 
                           tableId.name(), 
                           UUID.randomUUID());
    }
}